clusterloader2/pkg/prometheus/gce_windows_util.go

```diff
@@ -0,0 +1,112 @@
+)
+
+// gceNode is a struct storing gce node info
+type gceNode struct {
+	projectID string
+	zone      string
+	nodeName  string
+}
+
+func setUpWindowsNodeAndTemplate(k8sClient kubernetes.Interface, mapping map[string]interface{}) error {
+	// Get the gce windows node info
+	windowsNode, err := getGceWindowsNodeFromKubernetesService(k8sClient)
+	if err != nil {
+		return err
+	}
+	// Install the wmi exporter onto windows node
+	if err := installWmiExporter(k8sClient, windowsNode, mapping); err != nil {
+		return err
+	}
+	mapping["WINDOWS_NODE_NAME"] = windowsNode.nodeName
+	mapping["PROMETHEUS_SCRAPE_WINDOWS_NODES"] = true
+	return nil
+}
+
+func enableScrapingWindowsNode(mapping map[string]interface{}, clusterLoaderConfig *config.ClusterLoaderConfig) bool {
+	if windowsNodeTest, exists := mapping["WINDOWS_NODE_TEST"]; exists && windowsNodeTest.(bool) && clusterLoaderConfig.ClusterConfig.Provider == "gce" {
+		return true
+	}
+	return false
+}
+
+func installWmiExporter(k8sClient kubernetes.Interface, windowsNode *gceNode, mapping map[string]interface{}) error {
+	wmiExporterURL, ok := mapping["WMI_EXPORTER_URL"]
+	if !ok {
+		return fmt.Errorf("missing setting up wmi exporter download url")
+	}
+	wmiExporterCollectors, ok := mapping["WMI_EXPORTER_ENABLED_COLLECTORS"]
+	if !ok {
+		return fmt.Errorf("missing setting up wmi exporter enabled collectors")
+	}
+	installWmiCmdTemplate := `Start-Job -Name InstallWmiExporter -ScriptBlock {invoke-webrequest -uri %s -outfile C:\wmi_exporter.msi; Start-Process msiexec.exe -Wait -ArgumentList '/I C:\wmi_exporter.msi ENABLED_COLLECTORS=%s LISTEN_PORT=5000 /quiet'}; Wait-Job -Timeout 300 -Name InstallWmiExporter`
+	installWmiExporterCmd := fmt.Sprintf(installWmiCmdTemplate, wmiExporterURL, wmiExporterCollectors)
+	powershellCmd := fmt.Sprintf(`powershell.exe -Command "%s"`, installWmiExporterCmd)
+	if windowsNode == nil {
+		return fmt.Errorf("no windows nodes available to install wmi exporter")
+	}
+	klog.Infof("Installing wmi exporter onto projectId: %s, zone: %s, nodeName: %s with cmd: %s", windowsNode.projectID, windowsNode.zone, windowsNode.nodeName, installWmiExporterCmd)
+	cmd := exec.Command("gcloud", "compute", "ssh", "--project", windowsNode.projectID, "--zone", windowsNode.zone, windowsNode.nodeName, "--command", powershellCmd)
+	err := cmd.Run()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func getGceWindowsNodeFromKubernetesService(k8sClient kubernetes.Interface) (*gceNode, error) {
+	var nodeList *corev1.NodeList
+	f := func() error {
+		var err error
+		nodeList, err = k8sClient.CoreV1().Nodes().List(metav1.ListOptions{LabelSelector: "kubernetes.io/os=windows"})
+		return err
+	}
+
+	if err := client.RetryWithExponentialBackOff(client.RetryFunction(f)); err != nil {
+		return nil, err
+	}
+
+	if len(nodeList.Items) == 0 {
+		return nil, fmt.Errorf("no windows nodes available in kubernetes service")
+	}
+	// As we could scrape only one windows node as of now, so pick the first
```

Comment:
Can we add a TODO here to support more nodes in the future?

Reply:
Updated